File: src/Collega.Domain/IdeaFields/IdeaTypeFieldResolver.cs
using Collega.Domain.Fields;

namespace Collega.Domain.IdeaFields;

/// <summary>
/// The single source of truth for which User-Defined Fields an idea of a given type resolves to, and
/// whether each is required (SPEC/20-feature-idea-type-fields.md "Effective-field resolution"). Consumed
/// by the idea form, the value validator, and the detail projection so all three agree.
/// </summary>
public static class IdeaTypeFieldResolver
{
    /// <summary>
    /// Resolves the effective fields for <paramref name="ideaType"/>. The type's set resolves even when the
    /// type itself is soft-deleted — archival of the type does not change an existing idea's schema.
    /// </summary>
    /// <param name="ideaType">The idea's type (with its <see cref="IdeaType.Fields"/> links loaded).</param>
    /// <param name="activeOrgFieldDefinitions">
    /// The organization's active (non-soft-deleted) field definitions. A soft-deleted definition never
    /// appears in either branch, even if a <see cref="IdeaTypeFieldMode.Curated"/> link still references it.
    /// </param>
    public static IReadOnlyList<EffectiveField> ResolveEffectiveFields(
        IdeaType ideaType,
        IReadOnlyList<FieldDefinition> activeOrgFieldDefinitions)
    {
        ArgumentNullException.ThrowIfNull(ideaType);
        ArgumentNullException.ThrowIfNull(activeOrgFieldDefinitions);

        // Guard against a caller that passed archived definitions in — resolution never surfaces them.
        var activeById = activeOrgFieldDefinitions
            .Where(d => !d.IsDeleted)
            .ToDictionary(d => d.Id);

        if (ideaType.FieldMode == IdeaTypeFieldMode.AllActiveFields)
        {
            return activeById.Values
                .OrderBy(d => d.DisplayOrder)
                .ThenBy(d => d.Name, StringComparer.OrdinalIgnoreCase)
                .Select(d => new EffectiveField(d, d.IsRequired))
                .ToList();
        }

        return ideaType.Fields
            .Where(link => activeById.ContainsKey(link.FieldDefinitionId))
            .OrderBy(link => link.DisplayOrder)
            .ThenBy(link => activeById[link.FieldDefinitionId].Name, StringComparer.OrdinalIgnoreCase)
            .Select(link => new EffectiveField(activeById[link.FieldDefinitionId], link.IsRequired))
            .ToList();
    }
}

File: packages/domain/src/idea-fields/idea-type-field-resolver.ts
import { IdeaTypeFieldMode } from '../enums/index.js'
import type { FieldDefinition } from '../fields/index.js'
import type { EffectiveField } from './effective-field.js'
import type { IdeaType } from './idea-type.js'

/**
 * The single source of truth for which User-Defined Fields an idea of a given type resolves to,
 * and whether each is required (SPEC/20-feature-idea-type-fields.md "Effective-field
 * resolution"). Consumed by the idea form, the value validator (`validateFieldValues`), and the
 * detail projection so all three agree - this is the densest logic in the .NET Domain layer
 * (`IdeaTypeFieldResolver`), ported as a single pure function rather than a static class since
 * there is no state to hang a class off.
 *
 * Resolves even when `ideaType` itself is soft-deleted - archival of the type does not change an
 * existing idea's schema.
 *
 * @param ideaType The idea's type (with its `fields` links loaded).
 * @param activeOrgFieldDefinitions The organization's active (non-soft-deleted) field
 * definitions. A soft-deleted definition never appears in either branch below, even if a
 * `Curated` link still references it - the filter runs here too as a defence against a caller
 * that passed archived definitions in.
 */
export function resolveEffectiveFields(
  ideaType: IdeaType,
  activeOrgFieldDefinitions: readonly FieldDefinition[],
): readonly EffectiveField[] {
  const activeById = new Map(
    activeOrgFieldDefinitions
      .filter((definition) => !definition.isDeleted)
      .map((definition) => [definition.id, definition] as const),
  )

  if (ideaType.fieldMode === IdeaTypeFieldMode.AllActiveFields) {
    return [...activeById.values()]
      .sort((a, b) => a.displayOrder - b.displayOrder || compareIgnoreCase(a.name, b.name))
      .map((field) => ({ field, required: field.isRequired }))
  }

  return ideaType.fields
    .flatMap((link) => {
      const field = activeById.get(link.fieldDefinitionId)
      // A link naming an archived (or otherwise no-longer-active) definition is dropped here
      // rather than surfaced with a placeholder - `flatMap` returning `[]` filters it out without
      // a non-null assertion downstream.
      return field ? [{ link, field }] : []
    })
    .sort(
      (a, b) =>
        a.link.displayOrder - b.link.displayOrder || compareIgnoreCase(a.field.name, b.field.name),
    )
    .map(({ link, field }) => ({ field, required: link.isRequired }))
}

function compareIgnoreCase(a: string, b: string): number {
  const left = a.toLowerCase()
  const right = b.toLowerCase()
  if (left < right) {
    return -1
  }
  if (left > right) {
    return 1
  }
  return 0
}
